contrib/epee/include/net/abstract_tcp_server2.inl

```diff
@@ -242,10 +243,12 @@ PRAGMA_WARNING_DISABLE_VS(4355)
     boost::shared_ptr<connection<t_protocol_handler> >  back_connection_copy;
     LOG_TRACE_CC(context, "[sock " << socket_.native_handle() << "] release");
     CRITICAL_REGION_BEGIN(m_self_refs_lock);
-    CHECK_AND_ASSERT_MES(m_self_refs.size(), false, "[sock " << socket_.native_handle() << "] m_self_refs empty at connection<t_protocol_handler>::release() call");
-    //erasing from container without additional copy can cause start deleting object, including m_self_refs
-    back_connection_copy = m_self_refs.back();
-    m_self_refs.pop_back();
+    CHECK_AND_ASSERT_MES(m_reference_count, false, "[sock " << socket_.native_handle() << "] m_reference_count already at 0 at connection<t_protocol_handler>::release() call");
+    // is this the last reference?
+    if (--m_reference_count == 0) {
+        auto copy = m_self_ref;
+        m_self_ref.reset();
+    }
```

Comment:
This now destroys the last ref before the end of the function.

Reply:
This is indeed a behavioral change. In case the logger macro needs access to 'this' it could be a problem. I'll change this.